File: auto_editor/imports/fcp7.py
from __future__ import annotations

import xml.etree.ElementTree as ET
from fractions import Fraction
from typing import TYPE_CHECKING
from urllib.parse import unquote
from xml.etree.ElementTree import Element

from auto_editor.ffwrapper import FileInfo
from auto_editor.timeline import ASpace, Clip, Template, VSpace, v3

if TYPE_CHECKING:
    from auto_editor.utils.log import Log


SUPPORTED_EFFECTS = ("timeremap",)


def show(ele: Element, limit: int, depth: int = 0) -> None:
    print(
        f"{' ' * (depth * 4)}<{ele.tag} {ele.attrib}> {ele.text.strip() if ele.text is not None else ''}"
    )
    for child in ele:
        if isinstance(child, Element) and depth < limit:
            show(child, limit, depth + 1)


def read_filters(clipitem: Element, log: Log) -> float:
    for effect_tag in clipitem:
        if effect_tag.tag in {"enabled", "start", "end"}:
            continue
        if len(effect_tag) < 3:
            log.error("<effect> requires: <effectid> <name> and one <parameter>")
        for i, effects in enumerate(effect_tag):
            if i == 0 and effects.tag != "name":
                log.error("<effect>: <name> must be first tag")
            if i == 1 and effects.tag != "effectid":
                log.error("<effect>: <effectid> must be second tag")
                if effects.text not in SUPPORTED_EFFECTS:
                    log.error(f"`{effects.text}` is not a supported effect.")

            if i > 1:
                for j, parms in enumerate(effects):
                    if j == 0:
                        if parms.tag != "parameterid":
                            log.error("<parameter>: <parameterid> must be first tag")
                        if parms.text != "speed":
                            break

                    if j > 0 and parms.tag == "value":
                        if parms.text is None:
                            log.error("<value>: number required")
                        return float(parms.text) / 100

    return 1.0


def uri_to_path(uri: str) -> str:
    # Handle inputs like:
    # /Users/wyattblue/projects/auto-editor/example.mp4
    # file:///Users/wyattblue/projects/auto-editor/example.mp4
    # file:///C:/Users/WyattBlue/projects/auto-editor/example.mp4
    # file://localhost/Users/wyattblue/projects/auto-editor/example.mp4

    if uri.startswith("file://localhost/"):
        uri = uri[16:]
    elif uri.startswith("file://"):
        # Windows-style paths
        uri = uri[8:] if len(uri) > 8 and uri[9] == ":" else uri[7:]
    else:
        return uri
    return unquote(uri)


def read_tb_ntsc(tb: int, ntsc: bool) -> Fraction:
    if ntsc:
        if tb == 24:
            return Fraction(24000, 1001)
        if tb == 30:
            return Fraction(30000, 1001)
        if tb == 60:
            return Fraction(60000, 1001)
        return tb * Fraction(999, 1000)

    return Fraction(tb)


def fcp7_read_xml(path: str, log: Log) -> v3:
    def xml_bool(val: str) -> bool:
        if val == "TRUE":
            return True
        if val == "FALSE":
            return False
        raise TypeError("Value must be 'TRUE' or 'FALSE'")

    try:
        tree = ET.parse(path)
    except FileNotFoundError:
        log.error(f"Could not find '{path}'")

    root = tree.getroot()

    def parse(ele: Element, schema: dict) -> dict:
        new: dict = {}
        for key, val in schema.items():
            if isinstance(val, dict) and "__arr" in val:
                new[key] = []

        is_arr = False
        for child in ele:
            if child.tag not in schema:
                continue

            if schema[child.tag] is None:
                new[child.tag] = child
                continue

            if isinstance(schema[child.tag], dict):
                val = parse(child, schema[child.tag])
                is_arr = "__arr" in schema[child.tag]
            else:
                val = schema[child.tag](child.text)

            if child.tag in new:
                if not is_arr:
                    log.error(f"<{child.tag}> can only occur once")
                new[child.tag].append(val)
            else:
                new[child.tag] = [val] if is_arr else val

        return new

    def check(ele: Element, tag: str) -> None:
        if tag != ele.tag:
            log.error(f"Expected '{tag}' tag, got '{ele.tag}'")

    check(root, "xmeml")
    check(root[0], "sequence")
    result = parse(
        root[0],
        {
            "name": str,
            "duration": int,
            "rate": {
                "timebase": Fraction,
                "ntsc": xml_bool,
            },
            "media": None,
        },
    )

    tb = read_tb_ntsc(result["rate"]["timebase"], result["rate"]["ntsc"])
    av = parse(
        result["media"],
        {
            "video": None,
            "audio": None,
        },
    )

    sources: dict[str, FileInfo] = {}
    vobjs: VSpace = []
    aobjs: ASpace = []

    vclip_schema = {
        "format": {
            "samplecharacteristics": {
                "width": int,
                "height": int,
            },
        },
        "track": {
            "__arr": "",
            "clipitem": {
                "__arr": "",
                "start": int,
                "end": int,
                "in": int,
                "out": int,
                "file": None,
                "filter": None,
            },
        },
    }

    aclip_schema = {
        "format": {"samplecharacteristics": {"samplerate": int}},
        "track": {
            "__arr": "",
            "clipitem": {
                "__arr": "",
                "start": int,
                "end": int,
                "in": int,
                "out": int,
                "file": None,
                "filter": None,
            },
        },
    }

    sr = 48000
    res = (1920, 1080)

    if "video" in av:
        tracks = parse(av["video"], vclip_schema)

        if "format" in tracks:
            width = tracks["format"]["samplecharacteristics"]["width"]
            height = tracks["format"]["samplecharacteristics"]["height"]
            res = width, height

        for t, track in enumerate(tracks["track"]):
            if len(track["clipitem"]) > 0:
                vobjs.append([])
            for clipitem in track["clipitem"]:
                file_id = clipitem["file"].attrib["id"]
                if file_id not in sources:
                    fileobj = parse(clipitem["file"], {"pathurl": str})

                    if "pathurl" in fileobj:
                        sources[file_id] = FileInfo.init(
                            uri_to_path(fileobj["pathurl"]),
                            log,
                        )
                    else:
                        show(clipitem["file"], 3)
                        log.error(
                            f"'pathurl' child element not found in {clipitem['file'].tag}"
                        )

                if "filter" in clipitem:
                    speed = read_filters(clipitem["filter"], log)
                else:
                    speed = 1.0

                start = clipitem["start"]
                dur = clipitem["end"] - start
                offset = clipitem["in"]

                vobjs[t].append(
                    Clip(start, dur, sources[file_id], offset, stream=0, speed=speed)
                )

    if "audio" in av:
        tracks = parse(av["audio"], aclip_schema)
        if "format" in tracks:
            sr = tracks["format"]["samplecharacteristics"]["samplerate"]

        for t, track in enumerate(tracks["track"]):
            if len(track["clipitem"]) > 0:
                aobjs.append([])
            for clipitem in track["clipitem"]:
                file_id = clipitem["file"].attrib["id"]
                if file_id not in sources:
                    fileobj = parse(clipitem["file"], {"pathurl": str})
                    sources[file_id] = FileInfo.init(
                        uri_to_path(fileobj["pathurl"]), log
                    )

                if "filter" in clipitem:
                    speed = read_filters(clipitem["filter"], log)
                else:
                    speed = 1.0

                start = clipitem["start"]
                dur = clipitem["end"] - start
                offset = clipitem["in"]

                aobjs[t].append(
                    Clip(start, dur, sources[file_id], offset, stream=0, speed=speed)
                )

    T = Template.init(sources[next(iter(sources))], sr, res=res)
    return v3(tb, "#000", T, vobjs, aobjs, v1=None)

File: src/imports/fcp7.nim
import std/[algorithm, strutils, streams, tables, uri, xmlparser, xmltree]

import ../[action, av, ffmpeg, log, timeline]
import ../util/[color, rational]

func uriToPath(uri: string): string =
  var path = uri
  if path.startsWith("file://localhost/"):
    # Windows-style: file://localhost/C:/... -> C:/...
    if path.len > 18 and path[18] == ':':
      path = path[17..^1]
    else:
      path = path[16..^1]
  elif path.startsWith("file://"):
    # Windows-style: file:///C:/... -> C:/...
    if path.len > 9 and path[9] == ':':
      path = path[8..^1]
    else:
      # Unix-style: file:///path -> /path
      path = path[7..^1]
  else:
    return uri
  return decodeUrl(path, decodePlus = false)

func readTbNtsc(tb: int, ntsc: bool): AVRational =
  if ntsc:
    # ntsc scales the timebase by 1000/1001 (30 -> 29.97002997...), per the
    # chart in Apple's FCP7 XML docs; see the link in exports/fcp7.nim.
    return tb.int64 * AVRational(num: 1000, den: 1001)
  return AVRational(num: tb.cint, den: 1)

func num(s: string): (bool, float) =
  try: (true, parseFloat(s.strip()))
  except ValueError: (false, 0.0)

func readParam(parmNode: XmlNode): (string, seq[(float, float)]) =
  for c in parmNode:
    if c.kind != xnElement: continue
    case c.tag
    of "parameterid":
      result[0] = c.innerText.strip()
    of "value":
      let (ok, v) = num(c.innerText)
      if ok: result[1].add (0.0, v)
    of "keyframe":
      var timeOk, valOk = false
      var time, value = 0.0
      for kc in c:
        if kc.kind != xnElement: continue
        case kc.tag
        of "when": (timeOk, time) = num(kc.innerText)
        of "value": (valOk, value) = num(kc.innerText)
        else: discard
      if timeOk and valOk: result[1].add (time, value)
    else: discard

func addFixed(acts: var seq[Action], act: Action) =
  for a in acts:
    if a.kind == act.kind: return
  acts.add act

func lerpAt(pts: seq[(float, float)], time: float): float =
  if time <= pts[0][0]: return pts[0][1]
  for i in 1 ..< pts.len:
    let (t0, v0) = pts[i - 1]
    let (t1, v1) = pts[i]
    if time <= t1:
      if t1 <= t0: return v1
      return v0 + (v1 - v0) * (time - t0) / (t1 - t0)
  pts[^1][1]

func onGrid(pts: seq[(float, float)], srcIn, span: float): bool =
  for i, p in pts:
    if abs(p[0] - (srcIn + float(i) / float(pts.len - 1) * span)) > 0.5:
      return false
  true

func kfSeq(pts: seq[(float, float)], srcIn, srcOut, scale, lo, hi: float):
    seq[float32] =
  if pts.len == 0: return
  if pts.len == 1:
    if pts[0][1] * scale == 1.0: return
    return @[clamp(pts[0][1] * scale, lo, hi).float32]

  let ordered = pts.sorted
  let span = max(srcOut - srcIn - 1.0, 1.0)
  if ordered.len <= 255 and ordered.onGrid(srcIn, span):
    for p in ordered:
      result.add clamp(p[1] * scale, lo, hi).float32
    return

  let n = min(int(span) + 1, 255)
  for i in 0 ..< n:
    let time = srcIn + float(i) / float(n - 1) * span
    result.add clamp(ordered.lerpAt(time) * scale, lo, hi).float32

proc readFilters(filterNode: XmlNode, acts: var seq[Action],
    srcIn, srcOut: float) =
  for c in filterNode:
    if c.kind == xnElement and c.tag == "enabled" and
        c.innerText.strip() == "FALSE":
      return

  for effectTag in filterNode:
    if effectTag.kind != xnElement or effectTag.tag != "effect": continue

    var effectId = ""
    var params: Table[string, seq[(float, float)]]
    for effectChild in effectTag:
      if effectChild.kind != xnElement: continue
      if effectChild.tag == "effectid":
        effectId = effectChild.innerText.strip()
      elif effectChild.tag == "parameter":
        let (id, vals) = readParam(effectChild)
        if id != "" and vals.len > 0:
          params[id] = vals

    case effectId
    of "timeremap":
      if "speed" in params:
        let speed = params["speed"][0][1] / 100.0
        if speed > 0.0 and speed < 99999.0 and speed != 1.0:
          acts.addFixed Action(kind: actSpeed, val: speed.float32)
    of "opacity":
      let kf = kfSeq(params.getOrDefault("opacity"), srcIn, srcOut, 0.01, 0.0, 1.0)
      if kf.len > 0:
        acts.addFixed Action(kind: actOpacity, kf: kf)
    of "audiolevels":
      let kf = kfSeq(params.getOrDefault("level"), srcIn, srcOut, 1.0, 0.0, 3.98109)
      if kf.len > 0:
        acts.addFixed Action(kind: actVolume, kf: kf)
    else: discard

proc resolveFile(fileNode: XmlNode, sources: var Table[string, ptr string],
                 interner: var StringInterner): ptr string =
  let fileId = fileNode.attr("id")
  if fileId in sources:
    return sources[fileId]
  for fc in fileNode:
    if fc.kind != xnElement: continue
    if fc.tag == "pathurl":
      let filePath = uriToPath(fc.innerText.strip())
      let p = interner.intern(filePath)
      sources[fileId] = p
      return p
  return nil

proc getEffect(acts: seq[Action], effects: var seq[Actions]): uint32 =
  let actionGroup = newActions(acts)
  let idx = effects.find(actionGroup)
  if idx == -1:
    effects.add(actionGroup)
    return uint32(effects.len - 1)
  actionGroup.free()
  return uint32(idx)

proc parseTrack(trackNode: XmlNode, sources: var Table[string, ptr string],
                effects: var seq[Actions], interner: var StringInterner): seq[Clip] =
  for trackChild in trackNode:
    if trackChild.kind != xnElement: continue
    if trackChild.tag != "clipitem": continue

    var startVal = 0i64
    var endVal = 0i64
    var inVal = 0i64
    var outVal = -1i64
    var fileNode: XmlNode = nil
    var acts: seq[Action]
    var filterNodes: seq[XmlNode]

    for ci in trackChild:
      if ci.kind != xnElement: continue
      case ci.tag
      of "start": startVal = parseInt(ci.innerText.strip())
      of "end": endVal = parseInt(ci.innerText.strip())
      of "in": inVal = parseInt(ci.innerText.strip())
      of "out": outVal = parseInt(ci.innerText.strip())
      of "file": fileNode = ci
      of "filter": filterNodes.add ci
      else: discard

    if fileNode == nil: continue
    let srcPtr = resolveFile(fileNode, sources, interner)
    if srcPtr == nil: continue

    let dur = endVal - startVal
    if outVal <= inVal: outVal = inVal + dur
    for fn in filterNodes:
      readFilters(fn, acts, inVal.float, outVal.float)
    let e = getEffect(acts, effects)
    result.add Clip(src: srcPtr, start: startVal, dur: dur, offset: inVal,
      effects: e, stream: 0)

proc fcp7ReadXml*(path: string, interner: var StringInterner): v3 =
  let xmlContent = readFile(path)
  let stream = newStringStream(xmlContent)
  var parseErrors: seq[string] = @[]
  let root = parseXml(stream, path, parseErrors)
  stream.close()

  if parseErrors.len > 0:
    error "Failed to parse XML: " & parseErrors[0]

  if root == nil or root.tag != "xmeml":
    let tag = if root == nil: "nil" else: root.tag
    error "Expected 'xmeml' root tag, got '" & tag & "'"

  var seqNode: XmlNode = nil
  for child in root:
    if child.kind == xnElement and child.tag == "sequence":
      seqNode = child
      break

  if seqNode == nil:
    error "No <sequence> found in XML"

  var tb = AVRational(num: 30, den: 1)
  var mediaNode: XmlNode = nil

  for child in seqNode:
    if child.kind != xnElement: continue
    case child.tag
    of "rate":
      var timebase = 30
      var ntsc = false
      for rc in child:
        if rc.kind != xnElement: continue
        if rc.tag == "timebase":
          timebase = parseInt(rc.innerText.strip())
        elif rc.tag == "ntsc":
          ntsc = rc.innerText.strip() == "TRUE"
      tb = readTbNtsc(timebase, ntsc)
    of "media":
      mediaNode = child
    else: discard

  if mediaNode == nil:
    error "No <media> found in XML"

  var sr: cint = 48000
  var res = (1920'i32, 1080'i32)
  var sources: Table[string, ptr string]
  var effects: seq[Actions] = @[]
  var vobjs: seq[seq[Clip]] = @[]
  var aobjs: seq[seq[Clip]] = @[]

  for mediaChild in mediaNode:
    if mediaChild.kind != xnElement: continue
    case mediaChild.tag
    of "video":
      for vidChild in mediaChild:
        if vidChild.kind != xnElement: continue
        case vidChild.tag
        of "format":
          for fc in vidChild:
            if fc.kind != xnElement: continue
            if fc.tag == "samplecharacteristics":
              var w = 1920'i32
              var h = 1080'i32
              for sc in fc:
                if sc.kind != xnElement: continue
                if sc.tag == "width": w = parseInt(sc.innerText.strip()).int32
                elif sc.tag == "height": h = parseInt(sc.innerText.strip()).int32
              res = (w, h)
        of "track":
          let clips = parseTrack(vidChild, sources, effects, interner)
          if clips.len > 0:
            vobjs.add clips
        else: discard
    of "audio":
      for audChild in mediaChild:
        if audChild.kind != xnElement: continue
        case audChild.tag
        of "format":
          for fc in audChild:
            if fc.kind != xnElement: continue
            if fc.tag == "samplecharacteristics":
              for sc in fc:
                if sc.kind != xnElement: continue
                if sc.tag == "samplerate":
                  sr = parseInt(sc.innerText.strip()).cint
        of "track":
          let idx = audChild.attr("currentExplodedTrackIndex")
          if idx != "" and idx != "0": continue
          let clips = parseTrack(audChild, sources, effects, interner)
          if clips.len > 0:
            aobjs.add clips
        else: discard
    else: discard

  let bg = RGBColor(red: 0, green: 0, blue: 0)
  result = v3(tb: tb, bg: bg, sr: sr, res: res, v: vobjs, a: aobjs, effects: effects)
  result.templateFile = result.firstSource
  result.updateNumberOfSrc()
  result.layout = initLayout("stereo")
  while result.langs.len < result.v.len + result.a.len:
    result.langs.add ['u', 'n', 'd', '\0']
